### PR TITLE
crimson/os/seastore/lba_manager: do batch mapping allocs when remapping multiple mappings

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h [hunk 1]
```diff
@@ -320,6 +325,7 @@ class BtreeLBAManager : public LBAManager {
     std::vector<alloc_mapping_info_t> alloc_infos;
     for (auto &extent : extents) {
       alloc_infos.emplace_back(alloc_mapping_info_t{
+	extent->get_laddr(),
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h]: Sorry I'm not following here.

Looking at (should be) the only user: https://github.com/ceph/ceph/blob/adf4aca563c21106de1248a73fbef5a2bde5d7ab/src/crimson/os/seastore/transaction_manager.h#L359-L367

IIUC `extent->get_laddr()` here should always be NULL upon allocated by Cache, right?

So the new chance cannot be applied to any scenario?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h]: > So the new chance cannot be applied to any scenario?

Please see https://github.com/ceph/ceph/pull/57818#discussion_r1623926162

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h [hunk 2]
```diff
@@ -225,6 +225,7 @@ class BtreeLBAManager : public LBAManager {
 
 
   struct alloc_mapping_info_t {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h:227]: Let's add factory helper method to clarify how this is supposed to be constructed, e.g.
```c++
static alloc_mapping_info_t create_zero(len);
static alloc_mapping_info_t create_indirect(len, value/* intermediate_key */));
static alloc_mapping_info_t create_direct(len, value/* paddr */, crc, extent);
```

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 3]
```diff
@@ -396,13 +417,17 @@ BtreeLBAManager::_alloc_extents(
 		   c.trans, addr, total_len, hint, state.last_end);
 	    if (alloc_info.extent) {
 	      ceph_assert(alloc_info.val.is_paddr());
+	      assert(!alloc_info.extent->has_laddr() ||
+		alloc_info.extent->get_laddr() == iter.get_key());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:421]: How about this? :
```c++
if (alloc_info.extent->has_laddr()) {
  assert(alloc_info.key == alloc_info.extent->get_laddr());
  assert(alloc_info.key == iter.get_key());
} else {
  alloc_info.extent->set_laddr(iter.get_key());
}
```

**Review by cyx1231st — COMMENTED** [review]
Still looking.

Seems it is an intermediate step towards the direction to "iterate the lba-tree in one lookup to do update and insertions, or to do multiple removes".

In other words, I think when doing object data write (with the overwrite() implementations), the involved mappings (to update or insert) will be continuous.

If I'm correct above, the first commit of this PR (non-continuous mappings) will be finally deprecated, right?

**Comment by xxhdx1985126** [comment]
> If I'm correct above, the first commit of this PR (non-continuous mappings) will be finally deprecated, right?

If `iterating the lba-tree in one lookup to do update and insertions, or to do multiple removes` turns out to be feasible, Yes.

Actually, we are try to take 2 steps towards this goal:

1. implement "alloc_extents", "remove_mappings", "update_mappings", "move_mappings" and "clone_mappings", all of which are to do multiple same-type operations with one lba lookup;
2. implement the "batch modify" approach.

If the second step is not feasible, we have "batch same-type modify" ability at least.

**Comment by xxhdx1985126** [comment]
There's another use case for the `non-continuous` extents allocation:

At present, when doing clone/clone_range, we do two `ObjectDataHandler::clone_extents()` and one `ObjectDataHandler::do_removals()` to make the refcount of intermediate mappings correct. This is inefficient, we can do one `move_mappings` which move `direct mappings` to a new laddr space and one `ObjectDataHandler::clone_extents`. Since the mappings to be moved may not be continuous, we need the `non-continuous` extents allocation ability from LBAManager, so that the moves can be done with one lba lookup.

**Comment by cyx1231st** [comment]
Yeah, I think it's OK if the direction is aligned.

(unrelated to this PR below)

> implement "alloc_extents", "remove_mappings", "update_mappings", "move_mappings" and "clone_mappings", all of which are to do multiple same-type operations with one lba lookup;

Seems to me a dedicated "move_mappings" interface for user (ObjectDataHandler) might not be finally necessary, because it is always associated with a follow-up cloning. Probably the "LBA::clone_mappings" can hide&handle it internally when the "to-clone" mapping is direct. Merging "move_mappings" into "clone_mappings" also enables the optimization to directly set the ref-counter to 2 without another iteration.

Still, it should be fine to be in another step.

> Since the mappings to be moved may not be continuous, we need the non-continuous extents allocation ability from LBAManager, so that the moves can be done with one lba lookup.

> Meanwhile, OSDs' out-of-order recovery of clone objects also needs the move_mappings function

Ack, makes sense.

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 4]
```diff
@@ -307,10 +307,28 @@ BtreeLBAManager::_alloc_extents(
   std::vector<alloc_mapping_info_t> &alloc_infos,
   extent_ref_count_t refcount)
 {
+  ceph_assert(hint != L_ADDR_NULL);
   extent_len_t total_len = 0;
+#ifndef NDEBUG
+  bool laddr_null = (alloc_infos.front().key == L_ADDR_NULL);
+  laddr_t last_end = L_ADDR_NULL;
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:314]: Yes, the semantic meaning of the `BtreeLBAMapping::_alloc_extents()` is `to allocate multiple mappings within the range `hint~total_len`, the laddr of the first mapping may not be necessarily equal to the start of the range. Does this make sense to you?

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:314]: Yeah, let's also adjust the checks here to start with hint.

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc]: I moved this check out to `BtreeLBAMapping::remap_mappings()`, please take a look again, thanks:-)

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h:227]: Done:-)

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc]: Done.

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 5]
```diff
@@ -307,10 +307,28 @@ BtreeLBAManager::_alloc_extents(
...
+#ifndef NDEBUG
+  bool laddr_null = (alloc_infos.front().key == L_ADDR_NULL);
+  laddr_t last_end = hint;
   for (auto &info : alloc_infos) {
-    total_len += info.len;
+    assert((info.key == L_ADDR_NULL) == (laddr_null));
+    if (!laddr_null) {
+      assert(info.key >= last_end);
+      last_end = info.key + info.len;
+      last_end = alloc_infos.front().key + alloc_infos.front().len;
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:320]: Ah, sorry.

File: src/crimson/os/seastore/transaction_manager.h [hunk 6]
```diff
@@ -491,13 +492,15 @@ class TransactionManager : public ExtentCallbackInterface {
...
-	    extents.emplace_back(cache->alloc_remapped_extent<T>(
+	    auto extent = cache->alloc_remapped_extent<T>(
 	      t,
 	      remap_laddr,
 	      remap_paddr,
 	      remap_len,
 	      original_laddr,
-	      original_bptr));
+	      original_bptr);
+	    extent->set_laddr(remap_laddr);
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction_manager.h]: All done:-)

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 7]
```diff
@@ -402,7 +422,9 @@ BtreeLBAManager::_alloc_extents(
 	    rets.emplace_back(iter.get_pin(c));
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:422]: See another comment. I think in order to make sure the comment "// once assigned, the allocation to key must be exact and successful" is accurate, we should move the assertions
```c++
assert(mapping->get_key() == alloc_info.key);
assert(mapping->get_raw_val().get_laddr() ==
       alloc_info.val.get_laddr());
assert(mapping->get_length() == alloc_info.len);
```
into `_alloc_extents()`.

**Review by cyx1231st — APPROVED** [review]
Suggest some semantic revises, the implementation looks good.

**Comment by xxhdx1985126** [comment]
All done.

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
jenkins test windows

**Pull request merged**